### DISCUSSION TITLE
How to bypass MSW requests that are handled by the real backend?

### DISCUSSION THREAD

**Discussion by watch-janick** [mswjs/msw#1512] [Q&A]
Hello,

I'm wondering if it's possible to bypass any request that are handled by the real backend and only mock those who are missing?

My use-case is:
- I'm working on a front-end app using MSW to mock the back-end which is also under development.
- At some point, the back-end endpoint is available and should be used instead of MSW.
- I want the switch to be automatic without having to change my MSW implementation.

Is it possible?

Best,

Janick

**Comment by kettanaito** [comment]
Hey, @watch-janick. There are several options for how to approach this. 

First, you can define the handles only for the endpoints you wish to mock. MSW bypasses all requests by default, so those endpoints you haven't explicitly covered will hit the actual back-end. 

The second option, which is more applicable when you don't wish to manually manage handlers, you can always intercept all requests in MSW, perform a HEAD request to the original URL, check if the back-end returns anything (i.e. has the endpoint implemented), and return a mocked response only if it's missing. 

```js
export const handlers = [
  rest.all('*', async (req, res, ctx) => {
    const isImplemented = await ctx.fetch({
      ...req,
      method: 'HEAD',
    }).then((response) => response.ok).catch(() => false)
  
    if (isImplemented) {
      // If the lookup request to the backend was successful,
      // perform the intercepted request as-is.
      return req.passthrough()
    }
  
    // At this point, the backend doesn't know about this request.
    // Return a mocked response instead.
  }),
  rest.get('/greeting', (req, res, ctx) => {
    return res(ctx.text('hello'))
  })
]
```

Note that you won't be returning any mocked responses from the `rest.all()` handler. I'd highly suggest that since it'd allow you to keep your mocks granular vs packing all the mocked responses to a single handler. Since MSW resolves responses in sequence, if the `rest.all()` handler returns nothing, MSW will continue checking the subsequent handlers, like `rest.get('/greeting')`, allowing you to return mocked responses.

To recap: using the example above, that's what will happen:

1. If the `GET /greeting` _is_ implemented in the back-end, it will be performed as-is because `HEAD /greeting` will return a successful response;
2. If the `GET /greeting` _isn't_ implemented in the back-end, the `rest.all()` handled will return nothing, and MSW will hit the `rest.get('/greeting')` request handler, returning a mocked text response.
